src/test/java/org/cactoos/io/ZipTest.java

```diff
@@ -24,51 +24,113 @@
+                    )) {
+                        out.write(
+                            new BytesOf(
+                                new Zip(new Directory(folder))
+                            ).asBytes()
+                        );
+                    }
+                    try (ZipFile zipped = new ZipFile(
+                        new File(folder.getParentFile(), "empty.zip")
+                    )
+                    ) {
+                        return zipped.stream().map(
+                            ZipEntry::toString
+                        ).findFirst().get();
+                    }
+                }
+            ),
+            new ScalarHasValue<>(
+                "empty/"
+            )
+        ).affirm();
+    }
 
     @Test
-    public void zip() throws Exception {
-        final Path dir = this.folder.newFolder().toPath();
-        dir.resolve("x/y").toFile().mkdirs();
-        Files.write(dir.resolve("x/y/test"), "".getBytes());
-        try (ZipInputStream input = new ZipInputStream(
-            new Zip(
-                new Directory(dir)
-            ).stream()
-        )) {
-            int cnt = 0;
-            ZipEntry entry = input.getNextEntry();
-            while (entry != null) {
-                ++cnt;
-                entry = input.getNextEntry();
-            }
-            MatcherAssert.assertThat(
-                "Can't list files in a directory represented by a path",
-                cnt,
-                // @checkstyle MagicNumber (1 line)
-                Matchers.equalTo(4)
-            );
-        }
+    public void mustZipDirectory() throws Exception {
+        final String zipname = "abc.zip";
+        final File folder = this.temporal.newFolder("abc");
+        new File(folder, "A.txt").createNewFile();
+        new File(folder, "B").mkdir();
+        new File(
+            new File(
+                new Joined(File.separator, folder.getPath(), "B").asString()
+            ), "B.txt"
+        ).createNewFile();
+        new File(folder, "C").mkdir();
+        new Assertion<>(
+            "Must zip directory with the same directory structure",
+            new Sticky<>(
+                () -> {
+                    try (OutputStream out = new OutputStreamTo(
+                        this.temporal.newFile(zipname)
+                    )) {
+                        out.write(
+                            new BytesOf(
+                                new Zip(new Directory(folder))
+                            ).asBytes()
+                        );
+                    }
+                    try (ZipFile zipped = new ZipFile(
+                        new File(folder.getParentFile(), zipname)
+                    )
+                    ) {
+                        return new IterableOf<>(
+                            zipped.stream().map(
+                                e -> new TextOf(e.getName())
+                            ).collect(Collectors.toList())
```

Comment:
@fanifieiev can we get rid of stream and collect here?

Reply:
@victornoel I am not sure how to do without `Stream`, I actually tried, but could not make it work.
If you look inside `stream()` method of `ZipEntry`, you will see a lot of things happening there.

Reply:
@fanifieiev can't you wrap a stream into an `IterableOf` and then use `Mapped` on it for example? I think there should be something like that possible.